File: algopt/rebalancer/examples/website/solvers/local_search_examples.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Local Search Solver Examples
 *
 * This file demonstrates local search solver usage patterns shown in the
 * solver documentation.
 */

// example_start
#include "algopt/rebalancer/interface/ProblemSolver.h"
#include "algopt/rebalancer/interface/thrift/gen-cpp2/ProblemSpecs_types.h"
#include "algopt/rebalancer/interface/thrift/gen-cpp2/SolverSpecs_types.h"

#include <folly/executors/CPUThreadPoolExecutor.h>
#include <folly/init/Init.h>

#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace facebook::rebalancer::interface;

// Helper to set up a minimal problem for solver demonstrations
void setupMinimalProblem(ProblemSolver& solver) {
  solver.setObjectName("task");
  solver.setContainerName("host");

  std::map<std::string, std::vector<std::string>> assignment;
  assignment["host0"] = {"task0", "task1", "task2", "task3"};
  assignment["host1"] = {};
  assignment["host2"] = {};
  solver.setAssignment(assignment);

  std::map<std::string, double> cpu;
  cpu["task0"] = 1.0;
  cpu["task1"] = 2.0;
  cpu["task2"] = 1.5;
  cpu["task3"] = 2.5;
  solver.addObjectDimension("cpu", cpu);

  BalanceSpec balance;
  balance.name() = "balance-cpu";
  balance.scope() = "host";
  balance.dimension() = "cpu";
  balance.formula() = BalanceSpecFormula::LEGACY;
  balance.fixAverageToInitial() = true;
  solver.addGoal(balance, 1.0);
}

void quick_example() {
  /**
   * Quick example showing basic LocalSearchSolverSpec usage.
   */
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");

  setupMinimalProblem(solver);

  // quick_example_start
  LocalSearchSolverSpec spec;
  spec.moveTypeList()->push_back(
      ProblemSolver::makeMoveTypeSpec(SingleMoveTypeSpec()));
  spec.solveTime() = 30; // Stop after 30 seconds
  spec.stopAfterMoves() = 100000; // Or after 100K iterations

  solver.addSolver(spec);
  auto solution = solver.solve();

  // Check solve time
  std::cout << "Time: " << *solution.problemProfile()->solveSec() << "s\n";
  // quick_example_end
}

void fast_interactive() {
  /**
   * Fast interactive rebalancing for quick responses.
   */
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");

  // fast_interactive_start
  LocalSearchSolverSpec spec;
  spec.moveTypeList()->push_back(
      ProblemSolver::makeMoveTypeSpec(SingleMoveTypeSpec()));
  spec.solveTime() = 5; // 5 second limit

  solver.addSolver(spec);
  // fast_interactive_end
}

void production_rebalancing() {
  /**
   * Production rebalancing balancing speed and quality.
   */
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");

  // production_rebalancing_start
  LocalSearchSolverSpec spec;
  spec.moveTypeList()->push_back(
      ProblemSolver::makeMoveTypeSpec(SingleMoveTypeSpec()));
  spec.solveTime() = 60; // 1 minute
  spec.stopAfterMoves() = 500000; // 500K moves

  solver.addSolver(spec);
  // production_rebalancing_end
}

void offline_optimization() {
  /**
   * Offline optimization taking time for best solution.
   */
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");

  // offline_optimization_start
  LocalSearchSolverSpec spec;
  spec.moveTypeList()->push_back(
      ProblemSolver::makeMoveTypeSpec(SingleMoveTypeSpec()));
  spec.solveTime() = 3600; // 1 hour
  // No move limit - run until no improvements

  solver.addSolver(spec);
  // offline_optimization_end
}

void reproducible_results() {
  /**
   * Get reproducible results with fixed seed.
   */
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");

  // reproducible_results_start
  LocalSearchSolverSpec spec;
  spec.moveTypeList()->push_back(
      ProblemSolver::makeMoveTypeSpec(SingleMoveTypeSpec()));
  spec.randomSeed() = 42;

  solver.addSolver(spec);
  // reproducible_results_end
}

void increase_limits() {
  /**
   * Increase limits to improve solution quality.
   */
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");

  // increase_limits_start
  // Give more time
  LocalSearchSolverSpec spec;
  spec.moveTypeList()->push_back(
      ProblemSolver::makeMoveTypeSpec(SingleMoveTypeSpec()));
  spec.solveTime() = 300; // 5 minutes instead of 30 seconds
  spec.stopAfterMoves() = 1000000; // 1M moves instead of 100K

  solver.addSolver(spec);
  // increase_limits_end
}

int main(int argc, char* argv[]) {
  const folly::Init init(&argc, &argv);
  std::cout << "Running all Local Search Solver examples...\n\n";

  std::cout << "1. Quick Example...\n";
  quick_example();

  std::cout << "\n✓ All Local Search Solver examples completed successfully!\n";
  return 0;
}
// example_end

File: algopt/rebalancer/examples/website/solvers/local_search_examples.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Local Search Solver Examples

This file demonstrates local search solver usage patterns shown in the
solver documentation.
"""

# example_start
# pyre-strict
from algopt.rebalancer.interface.py_client.ProblemSolver import ProblemSolver
from rebalancer.interface.thrift.v2.ProblemSolver.thrift_types import SolverSpecs
from rebalancer.interface.thrift.v2.SolverSpecs.thrift_types import (
    LocalSearchSolverSpec,
)


def quick_example():
    """Quick example showing basic LocalSearchSolverSpec usage."""
    solver = ProblemSolver(service_name="example", service_scope="test")

    # quick_example_start
    solver.addSolver(
        SolverSpecs(
            localSearchSolverSpec=LocalSearchSolverSpec(
                timeLimitMs=30000,  # Stop after 30 seconds
                movesLimit=100000,  # Or after 100K iterations
            )
        )
    )

    solution = solver.solve()

    # Check termination reason
    print(f"Stopped because: {solution.profile.terminationReason}")
    print(f"Iterations: {solution.profile.iterations}")
    print(f"Time: {solution.profile.solveTime}ms")
    # quick_example_end


def fast_interactive():
    """Fast interactive rebalancing for quick responses."""
    solver = ProblemSolver(service_name="example", service_scope="test")

    # fast_interactive_start
    solver.addSolver(
        SolverSpecs(
            localSearchSolverSpec=LocalSearchSolverSpec(
                timeLimitMs=5000,  # 5 second limit
            )
        )
    )
    # fast_interactive_end


def production_rebalancing():
    """Production rebalancing balancing speed and quality."""
    solver = ProblemSolver(service_name="example", service_scope="test")

    # production_rebalancing_start
    solver.addSolver(
        SolverSpecs(
            localSearchSolverSpec=LocalSearchSolverSpec(
                timeLimitMs=60000,  # 1 minute
                movesLimit=500000,  # 500K moves
            )
        )
    )
    # production_rebalancing_end


def offline_optimization():
    """Offline optimization taking time for best solution."""
    solver = ProblemSolver(service_name="example", service_scope="test")

    # offline_optimization_start
    solver.addSolver(
        SolverSpecs(
            localSearchSolverSpec=LocalSearchSolverSpec(
                timeLimitMs=3600000,  # 1 hour
                # No move limit - run until no improvements
            )
        )
    )
    # offline_optimization_end


def reproducible_results():
    """Get reproducible results with fixed seed."""
    solver = ProblemSolver(service_name="example", service_scope="test")

    # reproducible_results_start
    solver.addSolver(SolverSpecs(localSearchSolverSpec=LocalSearchSolverSpec(seed=42)))
    # reproducible_results_end


def increase_limits():
    """Increase limits to improve solution quality."""
    solver = ProblemSolver(service_name="example", service_scope="test")

    # increase_limits_start
    # Give more time
    solver.addSolver(
        SolverSpecs(
            localSearchSolverSpec=LocalSearchSolverSpec(
                timeLimitMs=300000,  # 5 minutes instead of 30 seconds
                movesLimit=1000000,  # 1M moves instead of 100K
            )
        )
    )
    # increase_limits_end


if __name__ == "__main__":
    print("Running all Local Search Solver examples...\n")

    print("1. Quick Example...")
    quick_example()

    print("\n✓ All Local Search Solver examples completed successfully!")
# example_end
